SvPV_shrink_to_cur: include space for COW

richardleach opened the pull request:
The `SvPV_shrink_to_cur` macro shrinks an allocation to `SvCUR(sv) + 1`, which does not include an additional byte for Copy-On-Write (COW).

GH#22116 - a902d92a78262a0fd111789742f9c9e2a7fa2f42 - short-circuited constant folding on CONST OPs, as this should be unnecessary. However, Dave Mitchell noticed that it had the inadvertent effect of disabling COW on SVs holding UTF8 string literals (e.g. `"\x{100}abcd"`).

When a CONST OP is created, `Perl_ck_svconst` should mark its SV as being COW-able. But SVs built via `S_scan_const`, when that has called `SvPV_shrink_to_cur`, have resulting `SvLEN(sv)` values that fail the `SvCANCOW(sv)` test. Previously, constant folding had the effect of copying the literal into a buffer large enough for COW.

This commit modifies `SvPV_shrink_to_cur` so that it accounts for an additional byte for COW.

Additionally, this macro will no longer try to shrink the buffer below PERL_STRLEN_NEW_MIN, since that is likely to have no effect.

---------------------------------------------------------------------------------
* This set of changes does not require a perldelta entry.

richardleach committed:
SvPV_shrink_to_cur: don't be unrealistic, do include space for COW

The `SvPV_shrink_to_cur` macro shrinks an allocation to `SvCUR(sv) + 1`,
which does not include an additional byte for Copy-On-Write (COW).

GH#22116 - a902d92a78262a0fd111789742f9c9e2a7fa2f42 - short-circuited
constant folding on CONST OPs, as this should be unnecessary. However,
Dave Mitchell noticed that it had the inadvertent effect of disabling
COW on SVs holding UTF8 string literals (e.g. `"\x{100}abcd"`).

When a CONST OP is created, `Perl_ck_svconst` should mark its SV as
being COW-able. But SVs built via `S_scan_const`, when that has
called `SvPV_shrink_to_cur`, have resulting `SvLEN(sv)` values that
fail the `SvCANCOW(sv)` test. Previously, constant folding had the
effect of copying the literal into a buffer large enough for COW.

This commit modifies `SvPV_shrink_to_cur` so that it adds an
additional byte to allow for subsequent direct COWing.

The macro has also been modified to use the new `expected_size` macro,
so that reallocation will not be attempted if the resulting buffer
size is not going to be any smaller.

The intended saving is compared with SvLEN(sv), rather than SvCUR(sv),
as this also helps in cases such as SvCUR(sv) having previously been
set to zero and a size smaller than PERL_STRLEN_NEW_MIN is requested.
This will also enable some size checks at call sites to be simplified.

iabyn commented:
While this change seems sensible in general, it's changing the behaviour of an API macro: which makes me think it's not a suitable candidate for a last-minute change. Would it possible to have a more narrow fix which only affects utf8 literals, then apply the change in the PR after 5.42.0 is released?
Apart from that, I can confirm that this PR fixes the issue at hand, and also the performance regression which alerted me to the issue initially.

richardleach commented:
> Would it possible to have a more narrow fix which only affects utf8 literals, then apply the change in the PR after 5.42.0 is released?


Yes, the short circuit check in `Perl_op_convert_list` can be made to check to see if the CONST OP's SV has `IsCOW`, which should do it. I'll swap that for an assertion at the start of the next dev cycle.

xenu commented on sv.h:
```diff
@@ -1590,9 +1590,14 @@ L</C<SV_CHECK_THINKFIRST_COW_DROP>> before calling this.
+#define SvPV_shrink_to_cur(sv) STMT_START {              \
+                   const STRLEN _lEnGtH = SvCUR(sv) + 2; \
+                   if (PERL_STRLEN_NEW_MIN < _lEnGtH)    \
+                       SvPV_renew(sv, _lEnGtH);          \
```
This logic seems odd to me. Consider this pseudocode example:

```perl
my $foo = "aaa" x 1000; # CUR = 3000, LEN = 3002

$foo = "a"; # CUR = 1, LEN = 3002

SvPV_shrink_to_cur($foo); # will do nothing because CUR < PERL_STRLEN_NEW_MIN
```

I think it should be looking at SvLEN, not SvCUR.

richardleach replied:
True. Now that there's no rush to get this PR merged, I'll revisit it.

Similar SvCUR vs SvLEN considerations in https://github.com/Perl/perl5/pull/23111, where at least one of the callers does indeed set SvCUR to 0 before calling `S_invlist_trim`.

richardleach replied:
Will try to tie in ilmari's comment too: https://github.com/Perl/perl5/pull/23293#issuecomment-2885149691

richardleach replied:
Ok, the macro now checks `SvLEN` and the toke.c changes from https://github.com/Perl/perl5/pull/23293/commits/3dc75b6b5c6da243b4ce0041c34232327897a512 have been added as a second, now related, commit.

I haven't changed the callers in _pp_hot.c_ and _pp_sys.c_, since they are guarded by a `20` byte difference check and we haven't bottomed out if they need to be different and/or what a reasonable minimum saving looks like.

PS Not suggesting we think about using `malloc_good_size` et al here at this point. Or measuring the difference between "fastbins" and using that as the minimum saving.

bulk88 replied:
Not directly related to this commit, but on Threaded Win32 perl, all "malloc()" Perl core gets from Perl_safesysmalloc() end in 0x8 b/c of WinPerl VM "special-ness" b/c WinPerl adds a 3x PTRSIZE header to all mem blocks.

Unthreaded WinPerl (rarely used), will use whatever MS CRT gives it. On Win7 x64, the OS malloc header is always a fixed size of 8 bytes behind the PTR given to  the user. The user's mem block is always aligned to a unit of 16. Basically lowest hex digit is always 0x0. 16 byte alignment is because of SSE registers/SSE aka C lang FP math on x64. Below I have a realloc(len+1) loop, no WinPerl 3x PTRSIZE wrappers. It shows  "buckets" change every 0x10 bytes, on every length ending in 0x8. Since you hit the always `unaligned if SSE` malloc() header. `Ovrh`  means the `"cbOverhead"`  field as documented here https://learn.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-process_heap_entry . I personally don't know if I should "trust" that number.  I'm leaning on the side of trusting it, because I see the 4096 page/bucket change, at exactly 0x8 remaining of "overhead". So on windows minimum alloc on front end is, 3x PTRSIZE WinPerl header+ next length ending in 0x8. 
```
#ifndef PERL_STRLEN_NEW_MIN
#define PERL_STRLEN_NEW_MIN ((PTRSIZE == 4) ? 12 : 16)
#endif
```
And with a front end `SvLEN()` of 16 for all Perls
24+ 16 == 40, 40 == 0x28.
```
Realloc 000000000233F380 rq 40 0x28 ovrh 8 0x8 sum 48 0x30
Realloc 0000000000401830 rq 41 0x29 ovrh 23 0x17 sum 64 0x40
```
That was a bucket jump, So nothing wrong is happening with this rounding up math vs WinPerl here.
longer log:
```
Realloc 000000000041BB60 rq 1 0x1 ovrh 15 0xf sum 16 0x10
Realloc 000000000041BB70 rq 2 0x2 ovrh 14 0xe sum 16 0x10
Realloc 000000000041BB60 rq 3 0x3 ovrh 13 0xd sum 16 0x10
Realloc 000000000041BB70 rq 4 0x4 ovrh 12 0xc sum 16 0x10
Realloc 000000000041BB60 rq 5 0x5 ovrh 11 0xb sum 16 0x10
Realloc 000000000041BB70 rq 6 0x6 ovrh 10 0xa sum 16 0x10
Realloc 000000000041BB60 rq 7 0x7 ovrh 9 0x9 sum 16 0x10
Realloc 000000000041BB70 rq 8 0x8 ovrh 8 0x8 sum 16 0x10
Realloc 000000000041B4A0 rq 9 0x9 ovrh 23 0x17 sum 32 0x20
Realloc 000000000041B4E0 rq 10 0xa ovrh 22 0x16 sum 32 0x20
Realloc 000000000041B4A0 rq 11 0xb ovrh 21 0x15 sum 32 0x20
Realloc 000000000041B4E0 rq 12 0xc ovrh 20 0x14 sum 32 0x20
Realloc 000000000041B4A0 rq 13 0xd ovrh 19 0x13 sum 32 0x20
Realloc 000000000041B4E0 rq 14 0xe ovrh 18 0x12 sum 32 0x20
Realloc 000000000041B4A0 rq 15 0xf ovrh 17 0x11 sum 32 0x20
Realloc 000000000041B4E0 rq 16 0x10 ovrh 16 0x10 sum 32 0x20
Realloc 000000000041B4A0 rq 17 0x11 ovrh 15 0xf sum 32 0x20
Realloc 000000000041B4E0 rq 18 0x12 ovrh 14 0xe sum 32 0x20
Realloc 000000000041B4A0 rq 19 0x13 ovrh 13 0xd sum 32 0x20
Realloc 000000000041B4E0 rq 20 0x14 ovrh 12 0xc sum 32 0x20
Realloc 000000000041B4A0 rq 21 0x15 ovrh 11 0xb sum 32 0x20
Realloc 000000000041B4E0 rq 22 0x16 ovrh 10 0xa sum 32 0x20
Realloc 000000000041B4A0 rq 23 0x17 ovrh 9 0x9 sum 32 0x20
Realloc 000000000041B4E0 rq 24 0x18 ovrh 8 0x8 sum 32 0x20
Realloc 000000000233F3B0 rq 25 0x19 ovrh 23 0x17 sum 48 0x30
Realloc 000000000233F380 rq 26 0x1a ovrh 22 0x16 sum 48 0x30
Realloc 000000000233F3B0 rq 27 0x1b ovrh 21 0x15 sum 48 0x30
Realloc 000000000233F380 rq 28 0x1c ovrh 20 0x14 sum 48 0x30
Realloc 000000000233F3B0 rq 29 0x1d ovrh 19 0x13 sum 48 0x30
Realloc 000000000233F380 rq 30 0x1e ovrh 18 0x12 sum 48 0x30
Realloc 000000000233F3B0 rq 31 0x1f ovrh 17 0x11 sum 48 0x30
Realloc 000000000233F380 rq 32 0x20 ovrh 16 0x10 sum 48 0x30
Realloc 000000000233F3B0 rq 33 0x21 ovrh 15 0xf sum 48 0x30
Realloc 000000000233F380 rq 34 0x22 ovrh 14 0xe sum 48 0x30
Realloc 000000000233F3B0 rq 35 0x23 ovrh 13 0xd sum 48 0x30
Realloc 000000000233F380 rq 36 0x24 ovrh 12 0xc sum 48 0x30
Realloc 000000000233F3B0 rq 37 0x25 ovrh 11 0xb sum 48 0x30
Realloc 000000000233F380 rq 38 0x26 ovrh 10 0xa sum 48 0x30
Realloc 000000000233F3B0 rq 39 0x27 ovrh 9 0x9 sum 48 0x30
Realloc 000000000233F380 rq 40 0x28 ovrh 8 0x8 sum 48 0x30
Realloc 0000000000401830 rq 41 0x29 ovrh 23 0x17 sum 64 0x40
Realloc 00000000004015B0 rq 42 0x2a ovrh 22 0x16 sum 64 0x40
Realloc 0000000000401830 rq 43 0x2b ovrh 21 0x15 sum 64 0x40
Realloc 00000000004015B0 rq 44 0x2c ovrh 20 0x14 sum 64 0x40
Realloc 0000000000401830 rq 45 0x2d ovrh 19 0x13 sum 64 0x40
Realloc 00000000004015B0 rq 46 0x2e ovrh 18 0x12 sum 64 0x40
Realloc 0000000000401830 rq 47 0x2f ovrh 17 0x11 sum 64 0x40
Realloc 00000000004015B0 rq 48 0x30 ovrh 16 0x10 sum 64 0x40
Realloc 0000000000401830 rq 49 0x31 ovrh 15 0xf sum 64 0x40
Realloc 00000000004015B0 rq 50 0x32 ovrh 14 0xe sum 64 0x40
Realloc 0000000000401830 rq 51 0x33 ovrh 13 0xd sum 64 0x40
Realloc 00000000004015B0 rq 52 0x34 ovrh 12 0xc sum 64 0x40
Realloc 0000000000401830 rq 53 0x35 ovrh 11 0xb sum 64 0x40
Realloc 00000000004015B0 rq 54 0x36 ovrh 10 0xa sum 64 0x40
Realloc 0000000000401830 rq 55 0x37 ovrh 9 0x9 sum 64 0x40
Realloc 00000000004015B0 rq 56 0x38 ovrh 8 0x8 sum 64 0x40
Realloc 0000000000439590 rq 57 0x39 ovrh 23 0x17 sum 80 0x50
Realloc 00000000004396D0 rq 58 0x3a ovrh 22 0x16 sum 80 0x50
Realloc 0000000000439590 rq 59 0x3b ovrh 21 0x15 sum 80 0x50
Realloc 00000000004396D0 rq 60 0x3c ovrh 20 0x14 sum 80 0x50
Realloc 0000000000439590 rq 61 0x3d ovrh 19 0x13 sum 80 0x50
Realloc 00000000004396D0 rq 62 0x3e ovrh 18 0x12 sum 80 0x50
Realloc 0000000000439590 rq 63 0x3f ovrh 17 0x11 sum 80 0x50
Realloc 00000000004396D0 rq 64 0x40 ovrh 16 0x10 sum 80 0x50
Realloc 0000000000439590 rq 65 0x41 ovrh 15 0xf sum 80 0x50
Realloc 00000000004396D0 rq 66 0x42 ovrh 14 0xe sum 80 0x50
Realloc 0000000000439590 rq 67 0x43 ovrh 13 0xd sum 80 0x50
Realloc 00000000004396D0 rq 68 0x44 ovrh 12 0xc sum 80 0x50
Realloc 0000000000439590 rq 69 0x45 ovrh 11 0xb sum 80 0x50
Realloc 00000000004396D0 rq 70 0x46 ovrh 10 0xa sum 80 0x50
Realloc 0000000000439590 rq 71 0x47 ovrh 9 0x9 sum 80 0x50
Realloc 00000000004396D0 rq 72 0x48 ovrh 8 0x8 sum 80 0x50
Realloc 00000000004380E0 rq 73 0x49 ovrh 23 0x17 sum 96 0x60
Realloc 0000000000438320 rq 74 0x4a ovrh 22 0x16 sum 96 0x60
Realloc 00000000004380E0 rq 75 0x4b ovrh 21 0x15 sum 96 0x60
Realloc 0000000000438320 rq 76 0x4c ovrh 20 0x14 sum 96 0x60
Realloc 00000000004380E0 rq 77 0x4d ovrh 19 0x13 sum 96 0x60
Realloc 0000000000438320 rq 78 0x4e ovrh 18 0x12 sum 96 0x60
Realloc 00000000004380E0 rq 79 0x4f ovrh 17 0x11 sum 96 0x60
Realloc 0000000000438320 rq 80 0x50 ovrh 16 0x10 sum 96 0x60
Realloc 00000000004380E0 rq 81 0x51 ovrh 15 0xf sum 96 0x60
Realloc 0000000000438320 rq 82 0x52 ovrh 14 0xe sum 96 0x60
Realloc 00000000004380E0 rq 83 0x53 ovrh 13 0xd sum 96 0x60
```

richardleach replied:
> PS Not suggesting we think about using `malloc_good_size` et al here at this point. Or measuring the difference between "fastbins" and using that as the minimum saving.

Ok, I am thinking about measuring malloc behaviour at compile time, so sticking the "do not merge" label on this just for the moment.

richardleach replied:
I've tweaked this PR to add an `expected_size` stub macro, which could be replaced with a proper malloc-aware function in the future. The macro takes a size argument, returns `PERL_STRLEN_NEW_MIN` if the argument was smaller, or otherwise returns the size rounded up to the nearest pointer boundary. This also simplifies the definition of `SvPV_shrink_to_cur`.

I've still to refine the commit messages and check a couple of things, but wanted to ask your opinions on the idea in general in the meantime. (If the idea is fine, I'll use it for the other related PRs too.)

richardleach committed:
Simplify SvPV_shrink_to_cur() usage in toke.c

The previous commit modified the definition of `SvPV_shrink_to_cur` so that
it does not attempt to reallocate to make unrealistic savings. That made
some of the condition checks in toke.c redundant, so this commit removes
those.

richardleach committed:
Add Perl_expected_size stub macro

Passed a number of bytes, this macro will return:
* PERL_STRLEN_NEW_MIN if that is larger than the argument value
* the argument value rounded up to PTRSIZE

The intention is to later convert `Perl_expected_size` into a
function that has some knowledge of the underlying malloc library's
behaviour and can return values more reflective of the actual
useable allocation size for the specified number of bytes.

This can aid in reducing unnecessary calls to realloc().

richardleach committed:
Perldelta for Perl_expected_size / SvPV_shrink_to_cur
